### ISSUE TITLE
Missing actions are not being detected by default; mock_modules: not mocking actions when action validation explicitly enabled

### ISSUE THREAD

**Issue by ianw** [ansible/ansible-lint#2283]
In https://review.opendev.org/c/zuul/zuul-jobs/+/782004 a small bug was introduced; it calls ```zuul_azure_storage_upload:``` but the library name is ```zuul_azure_upload.py```.

Unfortunately ansible-lint didn't pick this up.  I found this by accident when I was testing an older version of ansible-lint (4.3.7) that did pick it up.

I bisected this down to change 61c64641cc7f0e94c8907d6cd657e30e176e0941.  This says ```skip_action_validation``` is "usually not needed", but I feel like this is actually required because we are not getting any validation of invalid action names (the comment may have been accurate at the time, but something in Ansible has now changed too?)

So my next step was to turn this on ```skip_action_validation``` with our current testing, using the latest release 6.4.0

This seems to reveal another issue; ``mock_modules:`` are not working correctly. e.g. you can easily replicate with the following commands to clone zuul-jobs, turn on ```skip_action_validation``` and then run ansible-lint (note the ```.ansible-lint``` in ```zuul-jobs``` has the ```mock_modules:``` already)

```console
$ git clone https://opendev.org/zuul/zuul-jobs
Cloning into 'zuul-jobs'...
remote: Enumerating objects: 7821, done.
remote: Counting objects: 100% (7821/7821), done.
remote: Compressing objects: 100% (3201/3201), done.
remote: Total 16105 (delta 7093), reused 4620 (delta 4620), pack-reused 8284
Receiving objects: 100% (16105/16105), 2.46 MiB | 598.00 KiB/s, done.
Resolving deltas: 100% (8884/8884), done.

$ cd zuul-jobs/
$ echo "skip_action_validation: false" >> .ansible-lint 
$ cat .ansible-lint 
parseable: true
quiet: false
skip_list:
  - meta-no-info  # No 'galaxy_info' found
  - no-changed-when  # Commands should not change things if nothing needs doing
  - no-tabs  # Most files should not contain tabs
  - role-name  # Role name does not match ``^[a-z][a-z0-9_]+$`` pattern
  - fqcn-builtins  # It would probably be good to enforce this, but it's a lot
  - risky-file-permissions  # It would probably also good to enforce this if someone wants to look at them
  - schema  # This is a stricter form of schema per https://github.com/ansible/schemas.  might also be good if someone has motivation?
use_default_rules: true
verbosity: 1
mock_modules:
  - zuul_console
  - zuul_return
loop_var_prefix: zj_
# Enable rules that are disabled by default:
enable_list:
  - no-same-owner

# Local variables:
# mode: yaml
# End:
skip_action_validation: false

$ tox -e linters
...
linters installed: ansible==2.10.7,ansible-base==2.10.17,ansible-compat==2.2.0,ansible-core==2.13.2,ansible-lint==6.4.0,attrs==22.1.0,bracex==2.3.post1,cffi==1.15.1,commonmark==0.9.1,cryptography==37.0.4,enrich==1.2.7,flake8==5.0.4,iniconfig==1.1.1,Jinja2==3.1.2,jsonschema==4.9.1,MarkupSafe==2.1.1,mccabe==0.7.0,packaging==21.3,pathspec==0.9.0,pluggy==1.0.0,py==1.11.0,pycodestyle==2.9.1,pycparser==2.21,pyflakes==2.5.0,Pygments==2.12.0,pyparsing==3.0.9,pyrsistent==0.18.1,pytest==7.1.2,PyYAML==6.0,resolvelib==0.8.1,rich==12.5.1,ruamel.yaml==0.17.21,ruamel.yaml.clib==0.2.6,subprocess-tee==0.3.5,tomli==2.0.1,wcmatch==8.4,yamllint==1.27.1
...
WARNING  Listing 29 violation(s) that are fatal
roles/ara-report/tasks/main.yaml:61: parser-error: couldn't resolve module/action 'zuul_return'. This often indicates a misspelling, missing collection, or incorrect module path.
roles/dstat-graph/tasks/main.yaml:26: parser-error: couldn't resolve module/action 'zuul_return'. This often indicates a misspelling, missing collection, or incorrect module path.
<and so on>
```

I tried to bisect this, but I ended up with a bunch of changes that I couldn't make work ```ImportError: cannot import name 'get_app' from 'ansiblelint.app'```.  Eventually git told me ```The first bad commit could be any of:```

* 688c9fc3a4949341f17ccdaad2c0f0d044171e78
* c4cdd50bdb190f43ed661d4f8f2d30c9d8b0d883
* f22a702248f5fc815ca9d11e8f747cdda49e22ea
* 56093f7868c93cf901788a4f0facf15710a02c44
* cfa997bc1efaa789f5b5ecc7b21c21e571afffc6
* 150ef7aae249d1df86ee1f70c613cc6f2108f6ce
* d24c49a6ba485447c83d7c38c3b554b201464c9d
* 24dd54ab3a8b3dca0ca2c3e215d6a050a8d028b9
* f44e2316b1342abe03bf48e1e3d9cbeb5c836546
* 2898dd06505c303c9786c8726d4375da84f9d800

The ones that look suspicious in there seem to be around ```ansible-compat```

So I think we have two things going on here

1) ```mock_modules``` isn't working to mock out missing actions since one of the changes above. This is hidden because ```skip_action_validation``` is turned on by default; meaning we're not checking for missing actions.
2) Reconsider if ```skip_action_validation``` being turned on by default is the right thing to do, because it does not seem like whatever is supposed to flag invalid actions is working.

**Comment by ssbarnea** [comment]
Thanks for reporting this Ian, please try to use the linter from main branch as I know that I recently fixed an issue related to mocking, and this was not released.

BTW, looking at "linters installed: ansible==2.10.7,ansible-base==2.10.17,ansible-compat==2.2.0,ansible-core==2.13.2,ansible-lint==6.4.0" ... does worry me as I am almost sure these packages should never end-up being reported inside the same environment, being incompatible. -- ansible vs base vs core! mixing various versions,... you know it.

**Comment by ianw** [comment]
> Thanks for reporting this Ian, please try to use the linter from main branch as I know that I recently fixed an issue related to mocking, and this was not released.

hrm, I just tried with a current pull of main branch @ 09a817ae2e460e4991e6928e895087d0a621798d -- just installed into the tox env with ```.tox/linters/bin/pip install <path>``` and it still fails with ```roles/ara-report/tasks/main.yaml:61: parser-error: couldn't resolve module/action 'zuul_return'``` errors.

```console
$ .tox/linters/bin/ansible-lint --version
WARNING: PATH altered to include /tmp/zuul-jobs/.tox/linters/bin
ansible-lint 6.4.1.dev14 using ansible 2.10.17
```
> BTW, looking at "linters installed: ansible==2.10.7,ansible-base==2.10.17,ansible-compat==2.2.0,ansible-core==2.13.2,ansible-lint==6.4.0" ... does worry me as I am almost sure these packages should never end-up being reported inside the same environment, being incompatible. -- ansible vs base vs core! mixing various versions,... you know it.

hrm, we are trying to keep "ansible" in sync with the Ansible's that Zuul runs; so we are capping "ansible" in ```requirements.txt```.  I did think the idea was this package kept the others in sync, but I don't really know?

c.f. https://review.opendev.org/c/zuul/zuul-jobs/+/851334 and https://review.opendev.org/c/zuul/zuul-jobs/+/851289/7

**Comment by ssbarnea** [comment]
You question regarding versions should be answered by https://ansible-lint.readthedocs.io/en/latest/philosophy/#linter-requires-an-ansible-version-newer-than-what-i-use-in-production - shortly you don't need to use same version. I know for sure that at least some ansible upgrades were breaking it and the only solution was to manually remove old version before installing new.

The funny bit is that zuul-jobs repository is part of our "eco" pipeline, we do run ansible-lint on it on each incoming PR, just to check if it does introduce undesired regressions. If the mocking would stop working, we would clearly see that. See https://github.com/ansible/ansible-lint/blob/main/test/test_eco.py#L21

**Comment by ianw** [comment]
> You question regarding versions should be answered by https://ansible-lint.readthedocs.io/en/latest/philosophy/#linter-requires-an-ansible-version-newer-than-what-i-use-in-production - shortly you don't need to use same version. I know for sure that at least some ansible upgrades were breaking it and the only solution was to manually remove old version before installing new.

Theoretically, couldn't testing with a later Ansible accept something that would break on an earlier version?  Since this is tox, it is at least starting fresh each time.  Might be interactions between the "core" packages etc?

> The funny bit is that zuul-jobs repository is part of our "eco" pipeline, we do run ansible-lint on it on each incoming PR, 

Does this test with `skip_action_validation`?  It seems like the mocking is moot if that is not enabled?

**Comment by ssbarnea** [comment]
The chance of breakage is very small, considerably smaller than encountering some already fixed bugs that would affect the linter negatively (unrelated to running ansible in production). I am aware of multiple changes that we needed from Ansible which were not backported to older versions. The second issue is that we will not be able to provide any support for older versions of linter or ansible, for the linter we do not have maintenance branches, only latest release is supported.

TBH, I the `skip_action_validation` is so old that forgot about its existence. I will have to dig a little bit more on it. My impression is that was needed in order make the mocking work, but I think that we might be able to update the mocking to render the feature obsolete, especially as the more recent versions of the linter are unable to parse most rules if ansible-playbook syntax check is failing anyway. Shortly, I do see action validation as becoming mandatory in the near future.

I am now in the middle of a couple of two WIP changes for the linter but I will look deeper on the zuul-jobs right after these.

BTW, if you can ping me on chat (matrix/gchat?), I might want to check few things with you. That might helps us do better decisions regarding how we would implement different features and lower the chance of negatively impacting zuul-jobs repository.

**Comment by ssbarnea** [comment]
Closing as the report is quite old and not directly actionable by us. Still, a PR that would improve the situation will be more than apprecited.
